### PR TITLE
Add support for Forge servers

### LINKED ISSUE #7: Add support for Forge servers
This was supposed to be a PR, but I wasn't able to finish what I started...
Instead of wasting that progress, I'll share what I found as an enhancement proposal instead:

- The recommended and latest versions of the Forge server by Minecraft version can be found [here](https://files.minecraftforge.net/net/minecraftforge/forge/promotions_slim.json), in JSON.
    - However, instead of a list, it's a large object, so to get the latest version without the user specifying a Minecraft version you can probably sort by keys
    - Same version list that `itzg/docker-minecraft-server` uses
- Forge JARs are named (roughly) as `(Minecraft version)-(Forge version)`-installer.jar`
    - Downloads are through their Maven repo, with a URL prefix of `https://maven.minecraftforge.net/net/minecraftforge/forge/`
    - Then, for example, to download a 1.19.3 server with version 44.1.23, append `(Minecraft version)-(Forge version)/forge-(Minecraft version)-(Forge version)-installer.jar` to the URL.

```go
// pseudo-go-code
prefix := "https://maven.minecraftforge.net/net/minecraftforge/forge"
mc := "1.19.3"
forge := forge.getLatestVersion()

fmt.Sprintf("%s/%s-%s/forge-%s-%s-installer.jar", prefix, mc, forge, mc, forge)
// https://maven.minecraftforge.net/net/minecraftforge/forge/1.19.3-44.1.23/forge-1.19.3-44.1.23-installer.jar
```

- The installer JAR can install the server (optionally with a GUI, if no arguments are supplied) like so: `forge-installer.jar --installServer [dir]`, if a directory is not specified it uses the current directory.
    - After installing, the installer JAR is safe to remove
- `run.bat` and `run.sh` are generated in the installation directory
    - The server itself can be run at `./libraries/net/minecraftforge/forge/(MC version)-(Forge version)-server.jar`, but the convenience scripts load arguments in the same directory which are required for Forge to work

Setting up a Forge server is relatively similar to other servers, EULA, `server.properties`, mods and all.

### PR DESCRIPTION
**Pull request by dkbednarczyk**
Aims to close #7.

- [x] Parse Minecraft/installer version
- [x] Download by Minecraft version
- [x] Download specific build
- [x] Add subcommand

Optional goals which were outlined in the original issue:
- [ ] Running installer with `--install` (or equivalent) flag
- [ ] Caching install logs

### CODE REVIEW

**Comment by talwat** [comment]
Looks good so far, and it aligns with the code for the rest of the jars.

**Comment by talwat** [comment]
I wonder if checksums are provided with the forge API, since it could help make it a little bit safer.

**Comment by dkbednarczyk** [comment]
Neither the official Forge website or the Github tags contain checksums unfortunately. I'm basing most of the code off [mc-image-helper](https://github.com/itzg/mc-image-helper/blob/master/src/main/java/me/itzg/helpers/forge/ForgeInstaller.java) and they don't check them there either.

**Comment by dkbednarczyk** [comment]
Trust me: my stab at parsing versions was ugly and I decided to fall back on [hashicorp/go-version](https://github.com/hashicorp/go-version), like in my [MRE](https://github.com/dxbednarczyk/forge-server). If needed, you can remove the dependency and try it yourself.
[Simplification is still needed.](https://codebeat.co/projects/github-com-talwat-pap-main/pull_requests/660750)

However, I think all of the major objectives are done.

**Comment by talwat** [comment]
Yeah, looks good. I don't really understand why you need to parse versions primarily because I don't use the forge downloads api but if you could just quickly explain it would be awesome. Also, if the dependency is needed and hard to implement then it's better to use an external dependency.

**Comment by talwat** [comment]
Also, could there be a flag to use a specific build/version?

**Comment by dkbednarczyk** [comment]
> I don't really understand why you need to parse versions primarily because I don't use the forge downloads api but if you could just quickly explain it would be awesome.

There isn't really one API for downloading... To find the latest version available you need to [sift through the keys of a `map[string]string` to find it](https://github.com/talwat/pap/pull/12/files#diff-788108843d60d0033145e4e66ef79be6e4bc1b64224c53b3f74af7cc11a10674R61), and just sorting by the key value itself doesn't produce a correctly sorted list of versions. `hashicorp/go-version` is helpful because it can turn a version string like `1.14.2-latest` into a struct

> Also, could there be a flag to use a specific build/version?

On it.

**Comment by talwat** [comment]
> There isn't really one API for downloading...

Ouch, that sounds painful.

**Comment by dkbednarczyk** [comment]
image
...what?

**Comment by dkbednarczyk** [comment]
Should be done. Lots of dumb mistakes on my part leading to extra, frankly unnecessary commits.

**Comment by talwat** [comment]
Looks like `golangci-lint` is making a few complaints. Mostly boils down to improper formatting and the like, if you could just fix the warnings it would be great. Also, for the variable names, it's easier and more readable to use longer variable names instead of shortenings unless it's **really** obvious (like i for index, pkg for package, and so on)

**Comment by dkbednarczyk** [comment]
Codebeat issues are all about assignment, not something I think is worth fixing. I think I caught all of the issues `golangci-lint` was complaining about (outside of test files, in which it complained about line length in long strings).

**Comment by talwat** [comment]
Yeah, there's also a couple of issues about white space and variable names, if you'd like I could fix them later. Also before this gets merged I want to update the CI to also test forge.

**Comment by talwat** [comment]
A couple of things, one, the CI is still failing (golangci-lint) and also the new CI you wrote is also having issues. I think it's because it is trying to parse `latest` as a version. golangci-lint generally is warning about short variable names, and because we are in a day and age of autocompletion there isn't too much of a need for super short variable names so you can just make them a little longer for clarity.

**Comment by dkbednarczyk** [comment]
Added `//nolint:` tags to functions which contained false positives and updated many, if not most variable names to be more verbose

**Comment by talwat** [comment]
Just a couple of lint warnings:

```txt
  Error: string `latest` has 3 occurrences, make it a constant (goconst)
  Error: append to slice `minecraftVersions` with non-zero initialized length (makezero)
  Error: parameter name 'i' is too short for the scope of its usage (varnamelen)
  ```
  
  The first one is easily solvable because there's a constant in the `jarfiles` package called `Latest`. The second one I have never seen before, and the third one you can either change to `idx` or just ignore it with `//nolint`. Also, the issue of the tests trying to parse `latest` as a version.

**Comment by dkbednarczyk** [comment]
Seems like these errors are caused by three different versioning schemes they use... really dumb. I'll try to fix a few of these later, but most (if not all) [*currently supported*](https://forums.minecraftforge.net/topic/91712-supported-version-directory/) Minecraft/installer versions work fine.

**Comment by dkbednarczyk** [comment]
I can only pray to god that this works with the CI, all test cases I added (which include edge cases) work on my machine...

After that it is just lint, lint, lint

**Comment by talwat** [comment]
So now the normal CI works, but the unit tests don't. There's also a couple of lint warnings which have to do with incorrect function and variable casing.

**Comment by dkbednarczyk** [comment]
What a rollercoaster this PR is for just a 500 line diff... man am I sloppy.

File: internal/global/global.go
Comment (talwat) [inline internal/global/global.go:1]: For the default value here, just put "latest" or whatever is in main.go

File: internal/jarfiles/forge/api.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	if installer == "latest" {
+		return minecraft, getVersion(&promos, &minecraft, "latest")
+	}
+
+	return minecraft, InstallerVersion{
+		Version: installer,
+	}
+}
+
+//nolint:interfacer
```
Comment (talwat) [inline internal/jarfiles/forge/api.go:120]: Can add an explanation towards why this nolint is here? You can do that by adding a space and then a comment after the nolint:interfacer

File: internal/jarfiles/forge/version.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	var err error
+	minecraft.Major, err = strconv.Atoi(splitVersion[0])
+	log.Error(err, "failed to parse major version")
+
+	minecraft.Minor, err = strconv.Atoi(splitVersion[1])
+	log.Error(err, "failed to parse minor version")
+
+	// to avoid magic numbers
+	lenMajorAndMinor := 2
```
Comment (talwat) [inline internal/jarfiles/forge/version.go]: Could you make this a const? Also, usually comments in pap are capitalised.

**Review by talwat — COMMENTED** [review]
Almost ready, just a few more finishing touches todo here and there.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -334,6 +334,14 @@ jobs:
            ./bin/pap -y -d download fabric --version 1.19.3 --snapshot
            ./bin/pap -y -d download fabric --version 20w22a
            ./bin/pap -y -d download fabric --version 1.16-pre8
+      - name: "Test forge downloading"
```
Comment (talwat) [inline .github/workflows/tests.yml:338]: Minor thing, but there should be a newline before this.

File: internal/cmd/downloadcmds/forge.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		global.MinecraftVersionInput,
+		global.ForgeInstallerVersion,
+		global.ForgeUseLatestInstaller,
+	)
+	surl := strings.Split(url, "/")
+
+	net.Download(
+		url,
+		"resolved forge-installer jarfile not found",
+		surl[len(surl)-1],
```
Comment (talwat) [inline internal/cmd/downloadcmds/forge.go]: Maybe instead just make it something like `forge.jar` or `forge-installer.jar`.

File: internal/jarfiles/forge/api.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/talwat/pap/internal/net"
+)
+
+func BuildURL(minecraft *MinecraftVersion, installer *InstallerVersion) string {
+	var returnURL string
+
+	returnURL += "https://maven.minecraftforge.net/net/minecraftforge/forge"
+	returnURL += fmt.Sprintf("/%s-%s", minecraft.String(), installer.Version)
+
+	if minecraft.Minor == 8 || minecraft.Minor == 9 {
```
Comment (talwat) [inline internal/jarfiles/forge/api.go:17]: Maybe explain these edge cases in comments?

File: internal/jarfiles/forge/struct.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
+package forge
+
+import (
+	"fmt"
+)
+
+type MinecraftVersion struct {
```
Comment (talwat) [inline internal/jarfiles/forge/struct.go:10]: Maybe add some struct documentation?

File: internal/jarfiles/forge/api.go
Comment (talwat) [inline internal/jarfiles/forge/api.go:1]: Perhaps you could add some debug logs (log.Debug) which would include information that could be used to debug later if there are issues.

**Review by talwat — CHANGES_REQUESTED** [review]
Just a few more things todo with general maintainability. Sorry, these should have been included in the original review.

File: internal/cmd/downloadcmds/forge.go (hunk 4 above)
Comment (dkbednarczyk) [reply internal/cmd/downloadcmds/forge.go]: After running the installer, it creates a file which ends in `forge.jar`. I'd rather keep the `-installer` suffix to avoid confusion

Comment (talwat) [reply internal/cmd/downloadcmds/forge.go]: I mean instead of using the last part of the url, just use `forge-installer.jar`

Comment (dkbednarczyk) [reply internal/cmd/downloadcmds/forge.go]: Can we meet in the middle and agree on a naming scheme such as
`forge-1.19-latest-installer.jar`? Could cause conflicts down the road if every file is named the same thing.

**Comment by talwat** [comment]
LGTM 👍

**Review by talwat — APPROVED** [review]

**Comment by talwat** [comment]
For some reason the tests for purpur are failing, but other than that I think it's ready to be merged.

**Comment by talwat** [comment]
Also a couple of golangci-lint errors about whitespace.

**Comment by dkbednarczyk** [comment]
Seems like it's going to inevitably complain whether or not it's formatted properly or not, the linters are complaining against each other... I'd say it's probably good to merge (purpur still failing)

**Comment by talwat** [comment]
You also now need to format it. It would be helpful if you just installed golangci-lint and ran it before commiting so we wouldn't have to go back and fourth with the lint warnings.

**Comment by dkbednarczyk** [comment]
> You also now need to format it.

If I run `gofumpt -s`, it will remove the whitespace that another linter previously complained was missing about, making it impossible to appease the CI

**Comment by talwat** [comment]
Well that's strange. Then you can just ignore that warning.

**Comment by talwat** [comment]
I will hopefully find the time to find a fix for the golangci-lint issue, but everything else looks good.
